### PR TITLE
#796 Improved the screen recorder for creating more than 3 minutes long videos

### PR DESCRIPTION
**Pull request by SergKhram**
https://github.com/MarathonLabs/marathon/issues/796
As mentioned there I prepared the solution
If apiLevel >= 34 -> we use time from VideoConfiguration
else -> we divide to 180sec videos if timeLimit > 180

### CODE REVIEW

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/RemoteFileManager.kt [hunk 1]
```diff
@@ -44,7 +44,7 @@ class RemoteFileManager(private val device: AndroidDevice) {
     }
 
     companion object {
-        const val MAX_FILENAME = 255
+        const val MAX_FILENAME = 254 //we need 1 more char for fileNumber in case of using video attachment > 180 && apiLevel < 34
```
Comment (Malinskiy) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/RemoteFileManager.kt]: This constant is not specific for video files so it should never change. It's a detail about Android filesystem implementation. I suggest changing the logic of the code, not the constants here

File: core/src/main/kotlin/com/malinskiy/marathon/extension/StringExtensions.kt [hunk 2]
```diff
@@ -24,3 +24,5 @@ fun String.escape(): String {
 }
 
 val escapeRegex = "[^a-zA-Z0-9\\.\\#]".toRegex()
+
+fun String.addFileNumberForVideo(fileNumber: String) = "${this.split(".mp4")[0]}$fileNumber.mp4"
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/extension/StringExtensions.kt]: This looks like very specific thing for stringextensions. I suggest putting this somewhere closer to the video file creation

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt [hunk 3]
```diff
@@ -122,29 +124,56 @@ class ScreenRecorderTestBatchListener(
...
+        val localVideoFiles = mutableListOf<File>()
         val remoteFilePath = device.fileManager.remoteVideoForTest(test, testBatchId)
         val millis = measureTimeMillis {
-            device.safePullFile(remoteFilePath, localVideoFile.toString())
+            if(device.apiLevel >= 34 || videoConfiguration.timeLimit <= 180) {
+                val localVideoFile = fileManager.createFile(FileType.VIDEO, pool, device.toDeviceInfo(), test, testBatchId)
+                device.safePullFile(remoteFilePath, localVideoFile.toString())
+                localVideoFiles.add(localVideoFile)
+            } else {
+                for (i in 0 .. (videoConfiguration.timeLimit / 180)) {
```
Comment (Malinskiy) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: I suspect that the actual number of files depends on the time the test took, not the configuration limit, so the loop should take into account only the time passed or/and number of shell invocations with `screenrecord` binary

**Review by Malinskiy — CHANGES_REQUESTED** [review]

**Comment by Malinskiy** [comment]
Thanks for submitting this @SergKhram 

The idea is solid, I think it needs a little bit more work on the handling the number of files as well as the file naming. Your assumption about needing only 1 character holds true only if there are less than 10 video files. Supporting this feature means that marathon should handle even more than that, so likely the trimming should happen the same way as it does for other files to fit into the max file name length

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt (hunk 3 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: done

File: core/src/main/kotlin/com/malinskiy/marathon/extension/StringExtensions.kt (hunk 2 above)
Comment (SergKhram) [reply core/src/main/kotlin/com/malinskiy/marathon/extension/StringExtensions.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/RemoteFileManager.kt (hunk 1 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/RemoteFileManager.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt [hunk 4]
```diff
@@ -372,7 +374,27 @@ class AdamAndroidDevice(
...
+                    recordsCount++
+                }
+            }
+        }
+    }
+
+    private suspend fun startScreenRecorder(
+        remoteFilePath: String,
+        options: VideoConfiguration,
+        fileNumber: String = "",
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: Can be just `counter:: Int? = null`. "fileNumber" as `String looks a bit weird`

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt [hunk 5]
```diff
@@ -372,7 +374,27 @@ class AdamAndroidDevice(
         remoteFilePath: String,
         options: VideoConfiguration
     ) {
-        val screenRecorderCommand = options.toScreenRecorderCommand(remoteFilePath)
+        var secondsRemaining = TimeUnit.SECONDS.convert(options.timeLimit, options.timeLimitUnits)
+        if(secondsRemaining <= 180 || apiLevel >= 34 || !options.increasedTimeLimitFeatureEnabled) {
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: Condition inversion

`secondsRemaining > 180 && apiLevel < 34 && options.increasedTimeLimitFeatureEnabled` - is positive way

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt [hunk 6]
```diff
@@ -12,5 +12,6 @@ data class VideoConfiguration(
     @JsonProperty("height") val height: Int = 1280,
     @JsonProperty("bitrateMbps") val bitrateMbps: Int = 1,
     @JsonProperty("timeLimit") val timeLimit: Long = 180,
-    @JsonProperty("timeLimitUnits") val timeLimitUnits: TimeUnit = TimeUnit.SECONDS
+    @JsonProperty("timeLimitUnits") val timeLimitUnits: TimeUnit = TimeUnit.SECONDS,
+    @JsonProperty("increasedTimeLimitFeatureEnabled") val increasedTimeLimitFeatureEnabled: Boolean = false
```
Comment (Vacxe) [inline configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt]: `increasedTimeLimitFeatureEnabled` maybe `longVideoSupport`

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt [hunk 7]
```diff
@@ -372,7 +374,27 @@ class AdamAndroidDevice(
...
+            startScreenRecorder(remoteFilePath, options)
+        } else {
+            var recordsCount = 0
+            while(recordsCount == 0 || secondsRemaining>=180) {
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: `secondsRemaining>=180` formatting

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt [hunk 8]
```diff
@@ -372,7 +374,27 @@ class AdamAndroidDevice(
...
+                }
+            }
+        }
+    }
+
+    private suspend fun startScreenRecorder(
+        remoteFilePath: String,
+        options: VideoConfiguration,
+        fileNumber: String = "",
+        finallyBlock: (() -> Unit)? = null
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: `finallyBlock` - `recordFinished`

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/extension/ConfigurationExtensions.kt [hunk 9]
```diff
@@ -36,7 +37,7 @@ fun VideoConfiguration.toScreenRecorderCommand(remoteFilePath: String): String {
     if (timeLimit > 0) {
         sb.append("--time-limit ")
         var seconds = TimeUnit.SECONDS.convert(timeLimit, timeLimitUnits)
-        if (seconds > 180) {
+        if (seconds > 180 && ((device?.apiLevel ?: 0) < 34 || !increasedTimeLimitFeatureEnabled)) {
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/extension/ConfigurationExtensions.kt]: Condition inversion
`seconds > 180 && (device?.apiLevel ?: 0) >= 34 && increasedTimeLimitFeatureEnabled`

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt [hunk 10]
```diff
@@ -34,5 +34,14 @@ internal class ScreenRecorder(
...
         private val logger = MarathonLogging.logger("ScreenRecorder")
+        private const val VIDEO_SUFFIX_LENGTH = 41 // "-" + 36 chars UUID of testBatchId + ".mp4"
+        fun String.addFileNumberForVideo(fileNumber: String): String  {
+            val lastIndexForReplacement = this.length - VIDEO_SUFFIX_LENGTH
+            var firstIndexForReplacement = lastIndexForReplacement
+            if (this.length + fileNumber.length > 255) {
+                firstIndexForReplacement -= fileNumber.length
+            }
+            return this.replaceRange(firstIndexForReplacement, lastIndexForReplacement, fileNumber)
+        }
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt]: Can you add tests for this logic please

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt [hunk 11]
```diff
@@ -110,44 +129,74 @@ class ScreenRecorderTestBatchListener(
...
 
     /**
      * This can be called both when test times out and device unavailable
      */
-    private suspend fun pullLastBatchVideo(remoteFilePath: String) {
-        val localVideoFile = fileManager.createFile(FileType.VIDEO, pool, device.toDeviceInfo(), testBatchId = testBatchId)
+    private suspend fun pullLastBatchVideo(remoteFilePath: String, testDuration: Long) {
         val millis = measureTimeMillis {
-            device.safePullFile(remoteFilePath, localVideoFile.toString())
+            if(device.apiLevel >= 34 || videoConfiguration.timeLimit <= defaultTimeLimit || !videoConfiguration.increasedTimeLimitFeatureEnabled) {
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: Invert please

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt [hunk 12]
```diff
@@ -110,44 +129,74 @@ class ScreenRecorderTestBatchListener(
...
+            }
         }
         logger.trace { "Pulling finished in ${millis}ms $remoteFilePath " }
     }
 
-    private suspend fun removeRemoteVideo(remoteFilePath: String) {
+    private suspend fun removeRemoteVideo(remoteFilePath: String, testDuration: Long) {
         val millis = measureTimeMillis {
-            device.fileManager.removeRemotePath(remoteFilePath)
+            if(device.apiLevel >= 34 || videoConfiguration.timeLimit <= defaultTimeLimit || !videoConfiguration.increasedTimeLimitFeatureEnabled) {
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: Invert

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt [hunk 13]
```diff
@@ -110,44 +129,74 @@ class ScreenRecorderTestBatchListener(
...
     }
 
-    private suspend fun pullTestVideo(test: Test) {
-        val localVideoFile = fileManager.createFile(FileType.VIDEO, pool, device.toDeviceInfo(), test, testBatchId)
+    private suspend fun pullTestVideo(test: Test, testDuration: Long) {
+        val localVideoFiles = mutableListOf<File>()
         val remoteFilePath = device.fileManager.remoteVideoForTest(test, testBatchId)
         val millis = measureTimeMillis {
-            device.safePullFile(remoteFilePath, localVideoFile.toString())
+            if(device.apiLevel >= 34 || videoConfiguration.timeLimit <= defaultTimeLimit || !videoConfiguration.increasedTimeLimitFeatureEnabled) {
```
Comment (Vacxe) [inline vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: Invert

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/extension/ConfigurationExtensions.kt (hunk 9 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/extension/ConfigurationExtensions.kt]: @Vacxe  But if (seconds > 180 && (device?.apiLevel ?: 0) >= 34 && increasedTimeLimitFeatureEnabled) == true -> we don't need to trim video, we need to do it if:
1) (device?.apiLevel ?: 0) < 34 AND seconds > 180 (even we have enabled feature -> we can't create > 180 sec videos for devices with apiLevel < 34) OR
2) our feature is disabled AND seconds > 180 (it doesn't matter what the api version - I want to save the old behavior in case of disabled feature)

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt (hunk 4 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt (hunk 5 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: done

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt (hunk 6 above)
Comment (SergKhram) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt (hunk 7 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt (hunk 8 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt (hunk 10 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorder.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt (hunk 11 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt (hunk 12 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: done

File: vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt (hunk 13 above)
Comment (SergKhram) [reply vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/video/ScreenRecorderTestBatchListener.kt]: done

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt [hunk 14]
```diff
@@ -12,5 +12,6 @@ data class VideoConfiguration(
     @JsonProperty("height") val height: Int = 1280,
     @JsonProperty("bitrateMbps") val bitrateMbps: Int = 1,
     @JsonProperty("timeLimit") val timeLimit: Long = 180,
-    @JsonProperty("timeLimitUnits") val timeLimitUnits: TimeUnit = TimeUnit.SECONDS
+    @JsonProperty("timeLimitUnits") val timeLimitUnits: TimeUnit = TimeUnit.SECONDS,
+    @JsonProperty("longVideoSupport") val longVideoSupport: Boolean = false
```
Comment (Malinskiy) [inline configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt]: We should not ask for any input here from the user. Same as the video recording falling back to screenshots automatically. Please remove

File: docs/runner/android/configure.md [hunk 15]
```diff
@@ -312,7 +312,8 @@ screenshots or configure the recording parameters you can specify this as follow
 
 :::tip
 
-Android's `screenrecorder` doesn't support videos longer than 180 seconds
+Android's `screenrecorder` doesn't support videos longer than 180 seconds for apiVersion < 34. If you want to set timeLimit > 180 you can
```
Comment (Malinskiy) [inline docs/runner/android/configure.md]:
```suggestion
Android's `screenrecorder` doesn't support videos longer than 180 seconds for apiVersion < 34. For such devices marathon will automatically record as many video files as needed if the test takes longer than 180 seconds. Expect several missing frames between the files though
```

File: docs/static/demo/android/allure-results/environment.xml [hunk 16]
```diff
@@ -1 +1 @@
...
...true, relativeResultsDirectory=/files/allure-results, pathRoot=APP_DATA), timeoutConfiguration=TimeoutConfiguration(shell=PT20S, listFiles=PT20S, pushFile=PT1M, pushFolder=PT1M, pullFile=PT30S, uninstall=PT20S, install=PT1M, screenrecorder=PT3M20S, screencapturer=PT0.3S, socketIdleTimeout=PT30S, portForward=PT20S, boot=PT30S), fileSyncConfiguration=FileSyncConfiguration(pull=[FileSyncEntry(relativePath=files/allure-results, pathRoot=APP_DATA, aggregationMode=TEST_RUN), FileSyncEntry(relativePath=/files/allure-results, pathRoot=APP_DATA, aggregationMode=TEST_RUN), FileSyncEntry(relativePath=coverage, pathRoot=APP_DATA, aggregationMode=POOL)], push=[]), threadingConfiguration=ThreadingConfiguration(bootWaitingThreads=4, adbIoThreads=4), testParserConfiguration=RemoteTestParserConfiguration(instrumentationArgs={listener=com.malinskiy.adam.junit4.android.listener.TestAnnotationProducer}), testAccessConfiguration=TestAccessConfiguration(adb=true, grpc=true, console=true, consoleToken=), adbServers=[AdbEndpoint(host=127.0.0.1, port=5037)], disableWindowAnimation=true)</value></parameter><parameter><key>deviceInitializationTimeoutMillis</key><value>180000</value></parameter></environment>
```
Comment (Malinskiy) [inline docs/static/demo/android/allure-results/environment.xml]: Why is this changed?

**Review by Malinskiy — CHANGES_REQUESTED** [review]

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt (hunk 14 above)
Comment (SergKhram) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt]: @Malinskiy but it's the toggle to enable new behavior - how we can control it wo the toggle?

File: docs/static/demo/android/allure-results/environment.xml (hunk 16 above)
Comment (SergKhram) [reply docs/static/demo/android/allure-results/environment.xml]: added the longVideoSupport toggle to VideoConfiguration

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt (hunk 14 above)
Comment (Malinskiy) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt]: There is no reason to control this. There are only 2 cases here:
- users who disabled video will still have the same behavior
- users who enabled video will get the extended recording if their tests are longer than 3 minutes

I dont see any reason to have this behaviour behind a flag: enhancements that do not break curremt behaviors should not be configurable by users.

If this was a breaking change then yes it would be required, but this is not the case

File: docs/static/demo/android/allure-results/environment.xml (hunk 16 above)
Comment (Malinskiy) [reply docs/static/demo/android/allure-results/environment.xml]: As mentioned above we shouldnt add any flags here so just revert it back please

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt (hunk 14 above)
Comment (SergKhram) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/VideoConfiguration.kt]: done

File: docs/runner/android/configure.md (hunk 15 above)
Comment (SergKhram) [reply docs/runner/android/configure.md]: done

File: docs/static/demo/android/allure-results/environment.xml (hunk 16 above)
Comment (SergKhram) [reply docs/static/demo/android/allure-results/environment.xml]: done

File: docs/static/demo/android/allure-results/environment.xml [hunk 17]
```diff
@@ -1 +1 @@
...
```
Comment (Malinskiy) [inline docs/static/demo/android/allure-results/environment.xml]: As mentioned before, this should not be included since there are no changes to the actual allure data files. The only case where this should change is when there is a new allure report demo which is not the case here

Comment (SergKhram) [reply docs/static/demo/android/allure-results/environment.xml:1]: @Malinskiy I already revert all changes to master , you can check by compare two branches for this file, but I don't know the reason why we see it here , and we have a lot of merges of master to this branch and I can't just revert all to recreate the commit without this file

**Comment by SergKhram** [comment]
@Vacxe waiting for merge
